### DISCUSSION TITLE
How to determine if command is available

### DISCUSSION THREAD

**Discussion by sebthom** [chrisant996/clink#438] [Q&A] [answered]
Does clink provide a convenient Lua function to determine if a command is available, i.e. is on PATH?

Something like e.g. `Clink.isCommandPresent("gcc")`.

**Comment by chrisant996** [comment, marked as answer]
No, but it's easy to do it yourself.

Loop over `ipairs(string.explode(os.getenv("path"), ";"))` and use `os.isfile(path.join(dir_from_path, name_to_find)`.  If the input `name_to_find` doesn't include an extension, then you'll want to also loop over either a list of predefined extensions or `ipairs(string.explode(os.getenv("pathext"), ";"))` and try `os.isfile(name_to_find .. each_ext_from_list)`.

**Reply by sebthom** [reply]
Ok, will do this, thanks! Maybe this is still worth a function in e.g. the os module?

**Reply by chrisant996** [reply]
> Ok, will do this, thanks! Maybe this is still worth a function in e.g. the os module?

@sebthom Maybe. What's the usage scenario where the need arose?

The trouble is that there need to be many nuanced variations depending on specifically what the caller is trying to do.
- Looking for files, or directories, or both?
- Looking for hidden files, system files, readonly files, symbolic links, or only normal files?
- Looking for only hard-coded extensions `.bat`, `.cmd`, `.com`, `.exe`?  Or not `.com`?  Or the extensions in `%PATHEXT%`?  Or the hard-coded ones plus `%PATHEXT%`?
- Looking for only files that are on local drives, i.e. not remote drives, e.g. to avoid hangs/delays due to network slowness?
- (and some others)

Two of the main criteria for making built-in Clink Lua APIs are:
- Is the thing commonly needed?
- Can the thing be done reasonably without special built-in support?

If the answers are NO and YES then it's probably not a good candidate for making a built-in Lua API function.  Of course, exceptions can potentially be made, depending on additional context and information.

A couple of examples:
- The family of `path.` functions is an interesting example.  Those are all easy to write yourself, right?  Well, actually no they're not -- there are many subtle pitfalls that people commonly fall into when trying to write their own versions of those (check out the sources for some potentially surprising complexity).  And they are super-commonly needed.  So the answers are YES and NOT REALLY, which makes it a strong candidate to be a built-in Lua API.
- The `io.popen()` function is MEDIUM and IMPOSSIBLE, which is another strong candidate to be built-in.

**Reply by sebthom** [reply]
It is about dynamically parsing the help file of a command to extract completion candidates.

I am now executing the Windows command `which <cmd>` and if it returns a path only then I invoke the actual command.

The problem I have is if I simply run io.popen then I get `XYZ is not recognized as an internal or external command` errors displayed in the console. So the stderr is not captured.

**Reply by sebthom** [reply]
Apparently my issue is something else. If I register a flag called **--define** and do `<tab><tab>` after it I get the `is not recognized as an internal or external command`. If I rename the flag to **--define1** it works as expected.

![image](https://user-images.githubusercontent.com/426959/221997996-ac627d24-6b5f-4dac-a246-a8cc1daaca03.png)

vs

![image](https://user-images.githubusercontent.com/426959/221998092-e758a787-ba3f-4c85-b0d6-6b5d32b2782c.png)

Even if I remove any code that would launch the haxe command I still get the `'haxe' is not recognized as an internal or external command`. Looks like clink is trying to invoke it?

**Reply by sebthom** [reply]
Ok, mistery solved. I had an old lua script that also got loaded into clink and created a conflicting registration for the command. Once I removed it no more error messages appearing and no need for a `Clink.isCommandPresent("gcc")` function.

**Reply by chrisant996** [reply]
@sebthom a few thoughts:
- I think that `which` was a typo, and you meant the `where` command, right?
- The fact that `--define1` works sounds like there may also be a mistake in how the command line is being constructed.  You might want to check whether redirection is being appended as `">something"`, which would then turn into `... --define1>something`, which CMD will interpret as a command `... --define` and a redirection of `1>something`.  Generally speaking, it is much safer to put redirection as the _first_ thing on a command line.  E.g. `2>nul the_rest_of_the_command_line` or `>somefile the_rest_of_the_command_line`.  Putting redirection first also enables CMD to reliably parse quotes on the rest of the command line, when the command name itself is quoted (e.g. `2>nul "foo.exe" --flag "value"`).
- Is there some kind of diagnostic or debugging mechanism Clink could provide to make it easier to detect the old script and conflicting registration?  E.g. `Alt-1`,`Ctrl-X`,`Ctrl-Z` reports a bunch of diagnostic info.  Is there something in that output that helps discover the root cause, or is there something that could be added that would help?

**Reply by sebthom** [reply]
Yes I meant the where command. I am switching a lot between windows/linux.

I think what would help is if clink displays a warning if two argmatchers try to register for the same command.

**Reply by chrisant996** [reply]
> I think what would help is if clink displays a warning if two argmatchers try to register for the same command.

That is explicitly a feature, and enables enhancing and merging argmatchers.  It's documented in [clink.argmatcher()](https://chrisant996.github.io/clink/clink.html#clink.argmatcher).

I can make the `clink-diagnostics` command list the sources that have contributed to a merged argmatcher.

But a warning isn't feasible.  In fact, it used to display a warning.  That's what led to user reports alerting me that v1.x had broken the merging feature that had existed since cb07751451b6721203b6efaf834f7f88813800d6 in v0.4.  That led me to add merging to v1.3.11 and fix some bugs in the v0.4.9 implementation and add some minor enhancements (Martin hadn't gotten around to adding it to the v1.x rewrite yet, or maybe had decided to cut it, I don't know).

Merging has varying degrees of success, depending on what's being done.  But for at least a while, Cmder was even making use of that IIRC.  I've also seen various other instances of intentionally using merging to extend existing argmatchers.  There is only so much that can be done for merging, though, as it can inherently get very complicated very quickly.

**Reply by chrisant996** [reply]
I forgot mention that for `clink-diagnostics` to list argmatchers an argument of 2 or greater is needed, e.g. `Alt-2`,`Ctrl-X`,`Ctrl-Z`.

Btw, "merging" is a little ambiguous.  The following is just defining an argmatcher.  But it uses an approach that is indistinguishable from merging.  So `clink-diagnostics` will report this as being a merged argmatcher, even though that wouldn't have been the intent.  Not a problem, but it could look weird/confusing if any scripts use that kind of approach.

**my_foo_argmatcher.lua**
```lua
clink.argmatcher("foo"):addflags("--the", "--flags", "--for", "--foo", "--command")
clink.argmatcher("foo"):addarg({
    "some", "values", "for", "argument", "position", "ONE"
}) -- This is to illustrate that checking for immediately adjacent source lines isn't viable for detecting this sort of case.
clink.argmatcher("foo"):addarg("and", "values", "for", "argument", "position", "TWO")
```

**How merging gets reported by the `clink-diagnostics` command**
```plain
...
argmatchers:
  ...
  foo                            :  c:\wbin\clink\my_foo_argmatcher.lua:1
    merged with                  :  c:\wbin\clink\my_foo_argmatcher.lua:2
    merged with                  :  c:\wbin\clink\my_foo_argmatcher.lua:5
  ...
...
```

**Reply by sebthom** [reply]
I am confused about how to invoke the clink-diagnostics command pressing ctrl+x or ctrl+z as listed in the documentation does nothing.

Is it possible to determine from code if an argmatcher is already registered for a given command? Since I created completions for openssl and curl in my repo similar ones now also exist in vladimir-kotikov/clink-completions so I would like to register my completion defensive, meaning only to register them as long as no binding is there already.

**Reply by chrisant996** [reply]
> I am confused about how to invoke the clink-diagnostics command pressing ctrl+x or ctrl+z as listed in the documentation does nothing.

Press `Alt-2` then `Ctrl-X` then `Ctrl-Z`. Unless you've bound the sequence to something else. And assuming you aren't using vi mode.

> Is it possible to determine from code if an argmatcher is already registered for a given command? Since I created completions for openssl and curl in my repo similar ones now also exist in vladimir-kotikov/clink-completions so I would like to register my completion defensive, meaning only to register them as long as no binding is there already.

[clink.getargmatcher()](https://chrisant996.github.io/clink/clink.html#clink.getargmatcher)
